---
title: "Code for Analyses and Plots in Abumazen et al."
author: Nooran Abu Mazen and Andrew C. Doxey
output:
  html_document: default
  fig_crop: no
---

This is an R Markdown document containing code and plots for Abumazen et al. "Nasopharyngeal metatranscriptomics reveals host-pathogen signatures of pediatric sinusitis".

## Loading required R packages

```{r, message=FALSE, warning=FALSE}
library(reshape2)
library(pheatmap)
library(pROC)
library(RColorBrewer)
library(dplyr)
library(plyr)
library(vioplot)
library(Metrics)
library(DESeq2)
library(tximport)
library(EnhancedVolcano)
library(patchwork)
library(knitr)

```

## 1. Kraken2-based pathogen detection from RNA-seq data

### Loading data files

------------------------------------------------------------------------

Reading in the kraken2 data

```{r, message=FALSE, warning=FALSE}
tb = read.delim("d1-d5_fastp_dehost_krakenmerged_reads_normalized.txt",sep='\t',col.names=(c("X","Y","Z")))

matr = dcast(tb,Y ~ X,value.var = "Z")
rownames(matr) = matr[,1]
matr = matr[,-1]

for (i in 1:nrow(matr))
{	matr[i,which(is.na(matr[i,]))] = 0
}

matr = t(data.matrix(matr))

splitrownames = vector(length = nrow(matr))
for (i in 1:length(splitrownames))
{
	rownames(matr)[i] = strsplit(rownames(matr),"_")[[i]][1]
}

```

Reading in the metadata

```{r, message=FALSE, warning=FALSE}

metadata = read.csv("metadata.12.20.updated.csv",header=T)
metadata = metadata[which(metadata$batch != 1),] #removing samples from batch 1

shared = intersect(rownames(matr),metadata[,1])
matr = matr[match(shared,rownames(matr)),]
metadata = metadata[match(shared,metadata[,1]),]

```

### Bacterial predictions

------------------------------------------------------------------------

### Heatmap

```{r, message=FALSE, warning=FALSE, fig.width = 4, fig.height = 8}

annotation_row = metadata[,c(22,24,26)]
annotation_row = (annotation_row -2) * -1   #just to make it going 0 to 1
rownames(annotation_row) = metadata[,1]
colnames(annotation_row) = c("SPN", "HFLU", "MCAT")


cols = c("white",colorRampPalette((brewer.pal(n = 7, name =
  "YlGnBu")))(99))

ann_colors = list(
    SPN = c("white", "light gray"),
    HFLU = c("white", "light gray"),
    MCAT = c("white", "light gray")
)


pheatmap(log(matr[,c(grep("Moraxella cat", colnames(matr)),grep("Streptococcus pne", colnames(matr)),grep("Haemophilus inf", colnames(matr)))] + 1, 10),annotation_row=annotation_row[,c(2,1,3)],cluster_cols = F, show_rownames = FALSE,col=cols,annotation_colors = ann_colors)

```

#### Box Plots

```{r, message=FALSE, warning=FALSE, fig.width = 6}

  group = metadata[,"mcat"]
	group <- mapvalues(group, 
          from=c("2","1"), 
          to=c("-","+"))
	group = as.factor(group)

	abund = log(matr[,"Moraxella catarrhalis"] + 1,10)
  p = ggplot(data.frame(mcat = abund), aes(y = abund, x = group,fill=as.factor(group))) +
  geom_boxplot(outlier.shape=NA) +
	ggtitle("MCAT") +	 theme(plot.title = element_text(hjust = 0.5)) +
  labs(y = "RNA-seq abundance + 1\nlog10(rpm)", x = "Culture test") +
  scale_fill_discrete(name = 'Culture test for MCAT', labels = c('Negative', 'Positive')) +
  geom_point(pch = 21, position = position_jitterdodge())

  p
  
  group = metadata[,"spn"]
		group <- mapvalues(group, 
          from=c("2","1"), 
          to=c("-","+"))
	group = as.factor(group)

	abund = log(matr[,"Streptococcus pneumoniae"] + 1,10)
	
  p = ggplot(data.frame(spn = abund), aes(y = abund, x = group,fill=as.factor(group))) +
  geom_boxplot(outlier.shape=NA) +
	ggtitle("SPN") +	 theme(plot.title = element_text(hjust = 0.5)) +
  labs(y = "RNA-seq abundance + 1\nlog10(rpm)", x = "Culture test") + 
  scale_fill_discrete(name = 'Culture test for SPN', labels = c('Negative', 'Positive')) +
  geom_point(pch = 21, position = position_jitterdodge())

  p
  
  group = metadata[,"hflu"]
		group <- mapvalues(group, 
          from=c("2","1"), 
          to=c("-","+"))
	group = as.factor(group)

	abund = log(matr[,"Haemophilus influenzae"] + 1,10)
	 
  p = ggplot(data.frame(hflu = abund), aes(y = abund, x = group,fill=as.factor(group))) +
  geom_boxplot(outlier.shape=NA) +
	ggtitle("HFLU") +	 theme(plot.title = element_text(hjust = 0.5)) +
  labs(y = "RNA-seq abundance + 1\nlog10(rpm)", x = "Culture test") +
  scale_fill_discrete(name = 'Culture test for HFLU', labels = c('Negative', 'Positive')) +
  geom_point(pch = 21, position = position_jitterdodge())

  p
  


```

#### *M. catarrhalis* prediction accuracy

```{r, message=FALSE, warning=FALSE, fig.height = 4, fig.width = 4}

bacterial_detection_threshold = 1

values = unique(rev(sort(matr[,"Moraxella catarrhalis"])))
TPR = vector(length = length(values)) 
FPR = vector(length = length(values)) 
for (i in 1:length(values))

{ k = values[i] 
    matches = which(matr[,"Moraxella catarrhalis"] >= k)
    TPR[i] = length(intersect(matches,which(metadata[,"mcat"] == 1))) / length(which(metadata[,"mcat"] == 1))
    FPR[i] = length(intersect(matches,which(metadata[,"mcat"] == 2))) / length(which(metadata[,"mcat"] == 2)) 
} 


prediction = matr[,"Moraxella catarrhalis"]
response = metadata[,"mcat"]
response = (response -2) * -1
auc(response,prediction)
auc_mcat = auc(response,prediction)

plot(FPR,TPR,type="l",main=paste("MCAT","AUC = ",round(auc_mcat,2)))

matches = which(matr[,"Moraxella catarrhalis"] > bacterial_detection_threshold)
tpr0 = length(intersect(matches,which(metadata[,"mcat"] == 1))) / length(which(metadata[,"mcat"] == 1))
fpr0 = length(intersect(matches,which(metadata[,"mcat"] == 2))) / length(which(metadata[,"mcat"] == 2)) 
points(x = fpr0, y = tpr0,cex=2)

sens_mcat = tpr0 * 100
spec_mcat = 100-(fpr0 * 100)

# prints out the sensitivity and the specificity given the threshold defined above
paste("mcat: sensitivity = ", round(tpr0 * 100,2), "%, specificity(%) = ", round(100-(fpr0 * 100),2), "%", sep='')

```

#### *S. pneumoniae* prediction accuracy

```{r, message=FALSE, warning=FALSE, fig.height = 4, fig.width = 4}

values = unique(rev(sort(matr[,"Streptococcus pneumoniae"])))
TPR = vector(length = length(values)) 
FPR = vector(length = length(values)) 
for (i in 1:length(values))

{ k = values[i] 
    matches = which(matr[,"Streptococcus pneumoniae"] >= k)
    TPR[i] = length(intersect(matches,which(metadata[,"spn"] == 1))) / length(which(metadata[,"spn"] == 1))
    FPR[i] = length(intersect(matches,which(metadata[,"spn"] == 2))) / length(which(metadata[,"spn"] == 2)) 
} 

prediction = matr[,"Streptococcus pneumoniae"]
response = metadata[,"spn"]
response = (response -2) * -1
auc(response,prediction)
auc_spn = auc(response,prediction)

plot(FPR,TPR,type="l",main=paste("spn","AUC = ",round(auc_spn,2)))

matches = which(matr[,"Streptococcus pneumoniae"] > bacterial_detection_threshold)
tpr0 = length(intersect(matches,which(metadata[,"spn"] == 1))) / length(which(metadata[,"spn"] == 1))
fpr0 = length(intersect(matches,which(metadata[,"spn"] == 2))) / length(which(metadata[,"spn"] == 2)) 
points(x = fpr0, y = tpr0,cex=2)


# prints out the sensitivity and the specificity given the threshold defined above
paste("spn: sensitivity = ", round(tpr0 * 100,2), "%, specificity(%) = ", round(100-(fpr0 * 100),2), "%", sep='')

sens_spn = tpr0 * 100
spec_spn = 100-(fpr0 * 100)
```

#### *H. influenzae* prediction accuracy

```{r, message=FALSE, warning=FALSE, fig.height = 4, fig.width = 4}
values = unique(rev(sort(matr[,"Haemophilus influenzae"])))
TPR = vector(length = length(values)) 
FPR = vector(length = length(values)) 
for (i in 1:length(values))

{ k = values[i] 
    matches = which(matr[,"Haemophilus influenzae"] >= k)
    TPR[i] = length(intersect(matches,which(metadata[,"hflu"] == 1))) / length(which(metadata[,"hflu"] == 1))
    FPR[i] = length(intersect(matches,which(metadata[,"hflu"] == 2))) / length(which(metadata[,"hflu"] == 2)) 
} 

prediction = matr[,"Haemophilus influenzae"]
response = metadata[,"hflu"]
response = (response -2) * -1
auc(response,prediction)
auc_hflu = auc(response,prediction)

plot(FPR,TPR,type="l",main=paste("hflu","AUC = ",round(auc_hflu,2)))

matches = which(matr[,"Haemophilus influenzae"] > bacterial_detection_threshold)
tpr0 = length(intersect(matches,which(metadata[,"hflu"] == 1))) / length(which(metadata[,"hflu"] == 1))
fpr0 = length(intersect(matches,which(metadata[,"hflu"] == 2))) / length(which(metadata[,"hflu"] == 2)) 
points(x = fpr0, y = tpr0,cex=2)

# prints out the sensitivity and the specificity given the threshold defined above
paste("hflu: sensitivity = ", round(tpr0 * 100,2), "%, specificity(%) = ", round(100-(fpr0 * 100),2), "%", sep='')

sens_hflu = tpr0 * 100
spec_hflu = 100-(fpr0 * 100)


accuracies = cbind(sens = c(mcat = sens_mcat,spn = sens_spn,hflu = sens_hflu),spec = c(mcat = spec_mcat,spec_spn,spec_hflu),auc = c(auc_mcat, auc_spn, auc_hflu))

kable(data.frame(lapply(data.frame(accuracies), FUN=round,digits=2), row.names=c("MCAT", "SPN", "HFLU")))

```

### Exploring diversity within streptococcus, haemophilus, and moraxella genus across dataset

```{r, message=FALSE, warning=FALSE, fig.height=10, fig.width=7}

allstreptococcus = matr[,grep("Streptococcus",colnames(matr))]
pheatmap(t(log(allstreptococcus + 1,10)),fontsize_col=5)

allmoraxella = matr[,grep("Moraxella",colnames(matr))]
pheatmap(t(log(allmoraxella + 1,10)),fontsize_col=5)

allhaemophilus = matr[,grep("Haemophilus",colnames(matr))]
pheatmap(t(log(allhaemophilus + 1,10)),fontsize_col=5)

```

### Viral predictions

------------------------------------------------------------------------

#### Heatmap

```{r, message=FALSE, warning=FALSE, fig.height=6, fig.width=6}

viral_reads = matrix(ncol=12,nrow = nrow(matr))

colnames(viral_reads) = c("INFA","INFB","INFC","MPV","RSV","HRV","PIV1","PIV2","PIV3","PIV4","ADV","EVD68")

rownames(viral_reads) = rownames(matr)

viral_reads[,"INFA"] = matr[,"Influenza A virus"]

viral_reads[,"INFB"] = matr[,"Influenza B virus"]

viral_reads[,"INFC"] = matr[,"Influenza C virus"]

viral_reads[,"MPV"] = matr[,"Human metapneumovirus"]

viral_reads[,"RSV"] = matr[,"Respiratory syncytial virus"] + matr[,"Human orthopneumovirus"]

viral_reads[,"HRV"] = matr[,"Rhinovirus A"] + matr[,"Rhinovirus B"] + matr[,"Rhinovirus C"]

viral_reads[,"PIV1"] = matr[,"Human respirovirus 1"]

viral_reads[,"PIV2"] = matr[,"Human orthorubulavirus 2"]

viral_reads[,"PIV3"] = matr[,"Human respirovirus 3"]

viral_reads[,"PIV4"] = matr[,"Human orthorubulavirus 4"]

viral_reads[,"ADV"] = matr[,"Human mastadenovirus B"] + matr[,"Human mastadenovirus C"]

viral_reads[,"EVD68"] = matr[,"Enterovirus D"]


tomap = log(viral_reads + 1, 10)

annotation = metadata[,c(34,32,59,57,55,53,49,63,51,46,40,36)]
rownames(annotation) = metadata[,1]
colnames(annotation) = rev(colnames(tomap))

cols = c("white",colorRampPalette((brewer.pal(n = 7, name =
  "YlGnBu")))(99))

for (i in 1:ncol(annotation))
{
	annotation[,i] = as.factor(findInterval(annotation[,i],c(10,20,30,40)))

}

ann_colors = list(
    INFA = c("4" = "#F2F2F2", "3" = "#E6E6E6", "2" = "dark gray", "1" = "black"),
    INFB = c("4" = "#F2F2F2", "3" = "#E6E6E6", "2" = "dark gray", "1" = "black"),
    INFC = c("4" = "#F2F2F2", "3" = "#E6E6E6", "2" = "dark gray", "1" = "black"),
    MPV = c("4" = "#F2F2F2", "3" = "#E6E6E6", "2" = "dark gray", "1" = "black"),
    RSV = c("4" = "#F2F2F2", "3" = "#E6E6E6", "2" = "dark gray", "1" = "black"),
    HRV = c("4" = "#F2F2F2", "3" = "#E6E6E6", "2" = "dark gray", "1" = "black"),
    PIV1 = c("4" = "#F2F2F2", "3" = "#E6E6E6", "2" = "dark gray", "1" = "black"),
    PIV2 = c("4" = "#F2F2F2", "3" = "#E6E6E6", "2" = "dark gray", "1" = "black"),
    PIV3 = c("4" = "#F2F2F2", "3" = "#E6E6E6", "2" = "dark gray", "1" = "black"),
    PIV4 = c("4" = "#F2F2F2", "3" = "#E6E6E6", "2" = "dark gray", "1" = "black"),
    ADV = c("4" = "#F2F2F2", "3" = "#E6E6E6", "2" = "dark gray", "1" = "black"),
    EVD68 = c("4" = "#F2F2F2", "3" = "#E6E6E6", "2" = "dark gray", "1" = "black")
)
pheatmap(tomap,
         annotation_row=annotation,
         annotation_colors = ann_colors,
         cluster_cols = F,
         show_rownames = F,
		 col=cols
)

```

#### Box plots

```{r, message=FALSE, warning=FALSE, fig.height=8, fig.width=4}


viruses = c("INFA","INFB","INFC","MPV","RSV","HRV","PIV1","PIV2","PIV3","PIV4","ADV","EVD68")


ctthreshold = 35

annotation = metadata[,c(34,32,59,57,55,53,49,63,51,46,40,36)]
rownames(annotation) = metadata[,1]
colnames(annotation) = rev(colnames(tomap))

annotation[annotation >= ctthreshold] = NA

annotation[annotation < ctthreshold] = 1

annotation[is.na(annotation)] = 0

par(mfrow=c(4,3), mar=c(4, 4, 2.5, 1.5))

sens_scores = vector(length = 12) # for each virus
spec_scores = vector(length = 12) # for each virus

for (viralcount in 1:12)
{	
		virus = viruses[viralcount]

		values = unique(rev(sort(viral_reads[,virus])))
		TPR = vector(length = length(values)) 
		FPR = vector(length = length(values)) 
		for (i in 1:length(values))
		{	k = values[i] 
		    matches = which(viral_reads[,virus] >= k)
		    TPR[i] = length(intersect(matches,which(annotation[,virus] == 1))) / length(which(annotation[,virus] == 1))
		    FPR[i] = length(intersect(matches,which(annotation[,virus] == 0))) / length(which(annotation[,virus] == 0)) 
		} 
		#plot(FPR,TPR,type="l",main=virus)

		boxplot(tomap[which(annotation[,virus] == 0),virus],tomap[which(annotation[,virus] == 1),virus],main=virus)

		matches = which(viral_reads[,virus] > 0)
		tpr0 = length(intersect(matches,which(annotation[,virus] == 1))) / length(which(annotation[,virus] == 1))
		fpr0 = length(intersect(matches,which(annotation[,virus] == 0))) / length(which(annotation[,virus] == 0)) 

		# prints out the sensitivity and the specificity
		#print(c(virus,tpr0 * 100,100-(fpr0 * 100)))

		sens_scores[viralcount] = as.numeric(tpr0 * 100)
		spec_scores[viralcount] = as.numeric(100 - (fpr0 * 100))
}

kable(cbind(virus = viruses,sens = round(sens_scores,2),spec = round(spec_scores,2)))

```

#### Optional: exploring prediction accuracy across multiple CT thresholds

```{r, message=FALSE, warning=FALSE, eval=FALSE}

avg_sensitivity = vector(length = 45)
avg_specificity = vector(length = 45)


for (count in 1:45) #this is the ct threshold

{
	ctthreshold = count

	annotation = metadata[,c(34,32,59,57,55,53,49,63,51,46,40,36)]
	rownames(annotation) = metadata[,1]
	colnames(annotation) = rev(colnames(tomap))

	annotation[annotation >= ctthreshold] = NA

	annotation[annotation < ctthreshold] = 1

	annotation[is.na(annotation)] = 0

	par(mfrow=c(4,3))

	sens_scores = vector(length = 12) # for each virus
	spec_scores = vector(length = 12) # for each virus

	for (viralcount in 1:12)
	{	
		virus = viruses[viralcount]

		values = unique(rev(sort(viral_reads[,virus])))
		TPR = vector(length = length(values)) 
		FPR = vector(length = length(values)) 
		for (i in 1:length(values))
		{	k = values[i] 
		    matches = which(viral_reads[,virus] >= k)
		    TPR[i] = length(intersect(matches,which(annotation[,virus] == 1))) / length(which(annotation[,virus] == 1))
		    FPR[i] = length(intersect(matches,which(annotation[,virus] == 0))) / length(which(annotation[,virus] == 0)) 
		} 
		plot(FPR,TPR,type="l",main=virus)

		matches = which(viral_reads[,virus] > 0)
		tpr0 = length(intersect(matches,which(annotation[,virus] == 1))) / length(which(annotation[,virus] == 1))
		fpr0 = length(intersect(matches,which(annotation[,virus] == 0))) / length(which(annotation[,virus] == 0)) 

		# prints out the sensitivity and the specificity
		print(c(virus,tpr0 * 100,100-(fpr0 * 100)))

		sens_scores[viralcount] = as.numeric(tpr0 * 100)
		spec_scores[viralcount] = as.numeric(100 - (fpr0 * 100))

	}

	print(paste(count,mean(sens_scores)))
	print(paste(count,mean(spec_scores)))

	avg_sensitivity[count] = mean(sens_scores)
	avg_specificity[count] = mean(spec_scores)

}

```

### Exploratory analysis of newly identified pathogens

------------------------------------------------------------------------

First, we will produce a heatmap of all kraken2-predicted taxa that contain "virus" in their organism names.

```{r, message=FALSE, warning=FALSE}
allViruses = matr[,grep("virus",colnames(matr))]
pheatmap(t(log(allViruses[,which(apply(allViruses,2,max) >= 10)] + 1,10)),fontsize_col=5)
```

### Calculation of total bacterial pathogen and viral abundance

------------------------------------------------------------------------

Here, we will calculate the total relative abundance of the three bacterial sinusitis pathogens (HFLU, MCAT, SPN), as well as the total relative abundance of respiratory viruses. This is done by summing their log10(RPM) values.

```{r, message=FALSE, warning=FALSE}

bacterial_pathogens = c("Acinetobacter baumannii","Acinetobacter johnsonii","Bordetella parapertussis","Chlamydia pneumoniae","Eikenella corrodens","Enterobacter cloacae","Haemophilus influenzae","Haemophilus parainfluenzae","Klebsiella aerogenes","Klebsiella pneumoniae","Moraxella catarrhalis","Neisseria meningitidis","Pseudomonas aeruginosa","Streptococcus anginosus","Staphylococcus aureus","Streptococcus agalactiae","Streptococcus pneumoniae","Streptococcus pyogenes")

viral_pathogens = c("Respiratory syncytial virus","Enterovirus D","Human coronavirus 229E","Human coronavirus HKU1","Betacoronavirus 1","Human coronavirus NL63","Human mastadenovirus B","Human mastadenovirus C","Human metapneumovirus","Human orthopneumovirus","Human orthorubulavirus 2","Human orthorubulavirus 4","Human respirovirus 1","Human respirovirus 3","Influenza A virus","Influenza B virus","Influenza C virus","Rhinovirus A","Rhinovirus B","Rhinovirus C")

three_bacteria = c("Haemophilus influenzae", "Moraxella catarrhalis", "Streptococcus pneumoniae")



bacPathAbundance = apply(matr[,three_bacteria],1,sum)  # sum of their individual RPMs

viralPathAbundance = apply(matr[,viral_pathogens],1,sum)   # sum of their individual RPMs


metadata = metadata %>% mutate(threebac_quantile = ntile(bacPathAbundance, 10), viral_quantile = ntile(viralPathAbundance, 10))


metadata$high_bac<- ifelse(metadata$threebac_quantile >=7, 1, 0)
metadata$high_viral<- ifelse(metadata$viral_quantile >=8, 1, 0)

metadata$high_pathogens <- ifelse(metadata$threebac_quantile >=7 & metadata$viral_quantile >=8,3, 
                           ifelse(metadata$threebac_quantile >=7, 2,
                           ifelse(metadata$viral_quantile >=8, 1, 0)))
```

## 2. Gene Expression Analyses

------------------------------------------------------------------------

### Loading data and formatting metadata

```{r, message=FALSE, warning=FALSE}

genesymbols = read.delim("gencode.v39.metadata.HGNC")

metadata = metadata[which(metadata$batch != 1),] #removing samples from batch 1
metadata[,"batch"] = as.factor(metadata[,"batch"])

metadata[,"bv_both_none"] = factor(metadata[,"bv_both_none"], labels = c("None detected", "Bacterial infection","Viral infection","Both"))

metadata[metadata[,"cum_pathogn"] > 1,"cum_pathogn"] = 1

metadata[,"cum_pathogn"] = as.factor(metadata[,"cum_pathogn"])

metadata[,"sex"] = as.factor(metadata[,"sex"])

metadata$age_scaled = scale(metadata$Age.in.years)

metadata[,"high_pathogens"] = factor(metadata[,"high_pathogens"],labels = c("Both low","High viral", "High bacterial", "Both high"))
```

### DESEQ2 pairwise comparison of bacteria-only vs virus-only group

```{r, message=FALSE, warning=FALSE}

onlyvirusonlybacteria_samples = metadata[which((metadata$bv_both_none == 1) | (metadata$bv_both_none == 2)),]

onlyvirusonlybacteria_samples = onlyvirusonlybacteria_samples %>% arrange(cum_pathogn)

subsetfiles <- file.path("C:/Users/noora/OneDrive - University of Waterloo/Doxey_lab_unibackup/salmon_output/alldata_without1095_1276", onlyvirusonlybacteria_samples$Filename, "/quant.sf")
subset_txi.salmon <- tximport(subsetfiles, type = "salmon", tx2gene = genesymbols)

dds <- DESeqDataSetFromTximport(subset_txi.salmon, onlyvirusonlybacteria_samples, ~ batch + sex + age_scaled + cum_pathogn)

dds <- DESeq(dds)

res <- results(dds)

summary(res)

pairwise_genes = which(res$padj < 0.001)

bac_upDEGs = intersect(which(res$padj < 0.001),which(res$log2FoldChange > 0))

viral_upDEGs = intersect(which(res$padj < 0.001),which(res$log2FoldChange < 0))


```

#### Number of DEGs detected:

There are `r length(bac_upDEGs)` bacterial upDEGs and `r length(viral_upDEGs)` viral upDEGs.

#### Volcano Plot

```{r, message=FALSE, warning=FALSE}
EnhancedVolcano(res,
  lab = rownames(res),
  x = 'log2FoldChange',
  y = 'pvalue',
  pCutoff = 0.001,
  FCcutoff = 0,
  xlim = c(-7, 7),
	ylim=c(-1,27),
  pointSize = 1.5,
  labSize = 2.5,
  title = 'DESeq2 results',
  subtitle = 'Differential expression',
  caption = 'p-value cutoff, 0.001',
  legendPosition = "none",
  legendLabSize = 14,
  col = c('grey30', 'light gray', 'royalblue', 'red2'),
  colAlpha = 0.9,
  drawConnectors = FALSE,
  widthConnectors = 0.5)

```

### Host-response / pathogen abundance correlations

Read in full set of files, including those with NO pathogens and BOTH (virus and bacteria)

```{r, message=FALSE, warning=FALSE}
files <- file.path("C:/Users/noora/OneDrive - University of Waterloo/Doxey_lab_unibackup/salmon_output/alldata_without1095_1276/", metadata$Filename, "/quant.sf")
txi.salmon <- tximport(files, type = "salmon", tx2gene = genesymbols)
pdds <- DESeqDataSetFromTximport(txi.salmon, metadata, ~ batch + sex + age_scaled + cum_pathogn)
prld <- varianceStabilizingTransformation(pdds, blind = TRUE)

```

### PCA plots

```{r, message=FALSE, warning=FALSE}
PCA <- plotPCA(prld[pairwise_genes,], intgroup = "bv_both_none") + stat_ellipse()
PCA

pdf("C:/Users/noora/OneDrive - University of Waterloo/Doxey_lab_unibackup/new paper/pca.pdf")
PCA <- plotPCA(prld[pairwise_genes,], intgroup = "high_pathogens") + stat_ellipse()
PCA
dev.off()
PCA
```

#### Jitter plots of expression levels for selected genes (previous biomarkers)

These are plots of host gene expression per clinical group

```{r, message=FALSE, warning=FALSE, fig.height=9, fig.width=7}

pathOrNot = as.numeric(metadata[,"bv_both_none"]) # using clinical categories

genelist = c("CXCL10", "PRF1", "IFI27", "CCL8", "PTGS2", "S100A9", "PLAUR", "TNFAIP3","IL1B")
for (i in 1:length(genelist)) {
  gene = genelist[i] 
  p = ggplot(data.frame(cbind(assay(prld)[gene,],pathOrNot)), aes(y = V1, x = as.factor(pathOrNot),fill=as.factor(pathOrNot))) +
  geom_boxplot(outlier.shape=NA) +
  ggtitle(gene) +
  theme(axis.title = element_blank(), legend.position="none") + 
	labs(y = "RNA-seq abundance\nlog10(rpm)", x = "Culture test") +
  geom_point(pch = 21, position = position_jitterdodge()) +
  scale_fill_discrete(labels=c("None detected", "Bacterial", "Viral", "Both")) +
  labs(fill = "Infection type")
  assign(paste0("p", i), p)
}
p1 + p2 + p3 + p4 + p5 + p6 + p7 + p8 + p9 + theme(legend.position="right") + plot_layout(nrow = 3)

```

Setting up data for sorted heatmaps

```{r, message=FALSE, warning=FALSE}

#setting up a temp matrix with Z-score scaled expression levels
temp = t(scale(t(assay(prld[,])))) 
temp[temp < -5] = -5
temp[temp > +5] = +5
temp[is.na(temp)] = 0

annotation_col = data.frame(
     Category = metadata$high_pathogens, 
		 Viral_pathogen_abundance = metadata[,"viral_quantile"],
     Bacterial_pathogen_abundance = metadata[,"threebac_quantile"],
     Symptom_severity_score = metadata[,"PRSS"],
     Days_with_cold = metadata[,"days_with_cold"],
     Infection = factor(metadata[,"bv_both_none"], labels = c("None","Bacterial", "Viral", "Both"))
     )
```

#### Viral sorted heatmap

```{r, message=FALSE, warning=FALSE, fig.width=10, fig.height=5}

heatmap <- pheatmap(temp[viral_upDEGs,order(viralPathAbundance)], 

	annotation_colors = list(Infection = c(None = "white", Bacterial = "light blue", Viral = "gold", Both = "black"),

    Resistance = c("No" = "white", "Yes" = "black")),
    annotation_col = annotation_col,
    show_colnames = F,
    show_rownames = F,
    border_color=NA,
    col = colorRampPalette( rev(brewer.pal(9, "RdBu")) )(55),
    #col = colorRampPalette(c("navy", "white", "red"))(50),
    fontsize_row=6,
    cluster_cols = F,
    cluster_rows = T,
    annotation_legend = TRUE)
```

#### Bacterial sorted heatmap

```{r, message=FALSE, warning=FALSE, fig.width=10, fig.height=5}

heatmap <- pheatmap(temp[bac_upDEGs,order(bacPathAbundance)], 

annotation_colors = list(Infection = c(None = "white", Bacterial = "light blue", Viral = "gold", Both = "black"),
    Resistance = c("No" = "green", "Yes" = "red")),
    annotation_col = annotation_col,
    show_colnames = F,
    show_rownames = F,
    border_color=NA,
    col = colorRampPalette( rev(brewer.pal(9, "RdBu")) )(55),
    #col = colorRampPalette(c("navy", "white", "red"))(50),
    fontsize_row=6,
    cluster_cols = F,
    cluster_rows = T,
    annotation_legend = TRUE)

```

#### Combined heatmap

```{r, message=FALSE, warning=FALSE, fig.width=10, fig.height=5}
heatmap1 <- pheatmap(temp[viral_upDEGs,order(metadata$high_pathogen)], 

	annotation_colors = list(Infection = c(None = "white", Bacterial = "light blue", Viral = "gold", Both = "black"),
    Resistance = c("No" = "white", "Yes" = "black")),
    annotation_col = annotation_col,
    show_colnames = F,
    show_rownames = F,
    border_color=NA,
    col = colorRampPalette( rev(brewer.pal(9, "RdBu")) )(55),
    #col = colorRampPalette(c("navy", "white", "red"))(50),
    fontsize_row=6,
    cluster_cols = F,
    cluster_rows = T,
    annotation_legend = TRUE,
	  main = "viral upDEGs")

heatmap2 <- pheatmap(temp[bac_upDEGs,order(metadata$high_pathogen)], 

	annotation_colors = list(Infection = c(None = "white", Bacterial = "light blue", Viral = "gold", Both = "black"),
    Resistance = c("No" = "white", "Yes" = "black")),
    annotation_col = annotation_col,
    show_colnames = F,
    show_rownames = F,
    border_color=NA,
    col = colorRampPalette( rev(brewer.pal(9, "RdBu")) )(55),
    #col = colorRampPalette(c("navy", "white", "red"))(50),
    fontsize_row=6,
    cluster_cols = F,
    cluster_rows = T,
    annotation_legend = TRUE,
	  main = "bacterial upDEGs")
```

### Jitter plots of host response scores vs pathogen abundance

```{r, message=FALSE, warning=FALSE}
viralResponseScore = apply(temp[viral_upDEGs,],2,mean)

bacterialResponseScore = apply(temp[bac_upDEGs,],2,mean)


group = as.numeric(metadata[,"threebac_quantile"])

  p = ggplot(data.frame(cbind(bacterialResponseScore,group, row.names=NULL)), aes(y = bacterialResponseScore, x = as.factor(group),fill=as.factor(group))) +
  geom_boxplot(outlier.shape=NA) +
  geom_point(pch = 21, position = position_jitterdodge()) +
  labs(fill = "Pathogen abundance") +
  labs(x = "Bacterial pathogen abundance (decile)", y = "Bacterial response score\n(Mean expression Z-score of bacterial upDEGs)") +
    theme(legend.position = "none")
  
  p


group = as.numeric(metadata[,"viral_quantile"])

  p = ggplot(data.frame(cbind(viralResponseScore,group, row.names=NULL)), aes(y = viralResponseScore, x = as.factor(group),fill=as.factor(group))) +
  geom_boxplot(outlier.shape=NA) +
  geom_point(pch = 21, position = position_jitterdodge()) +
  labs(fill = "Pathogen abundance") +
  labs(x = "Viral pathogen abundance (decile)", y = "Viral response score\n(Mean expression Z-score of viral upDEGs)") +
    theme(legend.position = "none")
  p


group = as.numeric(metadata[,"high_pathogens"])

  p = ggplot(data.frame(cbind(bacterialResponseScore,group, row.names=NULL)), aes(y = bacterialResponseScore, x = as.factor(group),fill=as.factor(group))) +
  geom_boxplot(outlier.shape=NA) +
  geom_point(pch = 21, position = position_jitterdodge()) +
  scale_fill_discrete(labels=c("Both low", "Viral high", "Bacterial high", "Both high")) +
  labs(fill = "Pathogen abundance") +
  labs(x = "Pathogen abundance (decile)", y = "Bacterial response score\n(Mean expression Z-score of bacterial upDEGs)") 
  p


group = as.numeric(metadata[,"high_pathogens"])

  p = ggplot(data.frame(cbind(viralResponseScore,group, row.names=NULL)), aes(y = viralResponseScore, x = as.factor(group),fill=as.factor(group))) +
  geom_boxplot(outlier.shape=NA) +
  geom_point(pch = 21, position = position_jitterdodge()) +
  scale_fill_discrete(labels=c("Both low", "Viral high", "Bacterial high", "Both high")) +
  labs(fill = "Pathogen abundance") +
  labs(x = "Pathogen abundance (decile)", y = "Viral response score\n(Mean expression Z-score of viral upDEGs)")
  
  
  p

```

### Scatter plots of group with high response for both bacterial and viral upDEGs

```{r, message=FALSE, warning=FALSE,fig.width=4, fig.height=5}
highresponse = which(metadata$high_pathogens == "Both high")

pdf("C:/Users/noora/OneDrive - University of Waterloo/Doxey_lab_unibackup/new paper/scatterhigh.pdf")
viralResponseScore = apply(temp[viral_upDEGs,highresponse],2,mean)
bacterialResponseScore = apply(temp[bac_upDEGs,highresponse],2,mean)
plot(viralResponseScore, bacterialResponseScore)
abline(lm(bacterialResponseScore ~ viralResponseScore))
summary(lm(bacterialResponseScore ~ viralResponseScore))  
dev.off()

pdf("C:/Users/noora/OneDrive - University of Waterloo/Doxey_lab_unibackup/new paper/scatterall.pdf")
viralResponseScore = apply(temp[viral_upDEGs,],2,mean)
bacterialResponseScore = apply(temp[bac_upDEGs,],2,mean)
plot(viralResponseScore, bacterialResponseScore)
abline(lm(bacterialResponseScore ~ viralResponseScore))
summary(lm(bacterialResponseScore ~ viralResponseScore))
dev.off()
```

#### Individual gene expression levels across groups

```{r, message=FALSE, warning=FALSE, fig.height=6, fig.width=10}
pathOrNot = as.numeric(metadata[,"high_pathogens"]) # using categories "Both low" "Viral high" "Bacterial high" "Both high"

genelist = c("PTGS2", "S100A9", "PLAUR", "TNFAIP3", "SERPINA2", "CXCL2","IL1B", "CXCL11","IFNL1","CXCL10", "PRF1", "IFI27", "CCL8","OAS2")
for (i in 1:length(genelist)) {
  gene = genelist[i] 
  p = ggplot(data.frame(cbind(assay(prld)[gene,],pathOrNot)), aes(y = V1, x = as.factor(pathOrNot),fill=as.factor(pathOrNot))) +
  geom_boxplot(outlier.shape=NA) +
  ggtitle(gene) +
  theme(axis.title = element_blank(), legend.position="none") +
  geom_point(pch = 21, position = position_jitterdodge()) +
  scale_fill_discrete(labels=c("None detected", "Bacterial", "Viral", "Both")) +
  labs(fill = "Infection type")
  assign(paste0("p", i), p)
}
p1 + p2 + p3 + p4 + p5 + p6 + p7 + p8 + p9 + p10 + p11 + p12 + p13 + p14 + plot_layout(nrow = 2)

```

#### Testing the specificity of the bacterial host response for SPN/HFLU/MCAT

Here, we will compute the correlation between the bacterial host response score and the total bacterial pathogen abundance (MCAT + SPN + HFLU), and compare this to 1000 randomly samples selections of three bacterial species from the data.

```{r, message=FALSE, warning=FALSE}
abundant_bacteria = setdiff(which(apply(matr,2,max) >= 10),grep("virus",colnames(matr)))
cors = vector(length = 1000)
for (i in 1:1000)
{

		randSetOfThree = sample(abundant_bacteria,3)
		threeBacAbundance = apply(matr[,randSetOfThree],1,sum)  # sum of their individual RPMs
		#ntile(threeBacAbundance, 10)
		cors[i] = cor(bacterialResponseScore,log(threeBacAbundance + 1,10))
}

hist(cors,xlim=c(-1,1),breaks=100)
abline(v = cor(bacterialResponseScore, log(bacPathAbundance + 1,10)))

```

#### ROC curves for predicting high/low pathogen abundance based on host response

Viral ROC

```{r, message=FALSE, warning=FALSE}

response = vector(length = nrow(metadata))
response[] = 0
response[which(metadata$high_pathogen == 1 | metadata$high_pathogen == 3)] = 1

prediction = viralResponseScore
bacPlot = plot.roc(roc(response,prediction))
#coords(bacPlot) #for a list of sensitivities and specificities

auc(response,prediction)

```

Bacterial ROC

```{r, message=FALSE, warning=FALSE}
response = vector(length = nrow(metadata))
response[] = 0
response[which(metadata$high_pathogen == 2 | metadata$high_pathogen == 3)] = 1

prediction = bacterialResponseScore
bacPlot = plot.roc(roc(response,prediction))
#coords(bacPlot) #for a list of sensitivities and specificities

auc(response,prediction)

```

```{r, message=FALSE, warning=FALSE}

```

```{r, message=FALSE, warning=FALSE}

```

```{r, message=FALSE, warning=FALSE}

```
